numpyro/util.py

```diff
@@ -197,26 +200,28 @@ def fori_collect(lower, upper, body_fun, init_val, transform=identity,
         collected along the leading axis of `np.ndarray` objects.
     """
     assert lower <= upper
-    collection_size = upper - lower if collection_size is None else collection_size
-    assert collection_size >= upper - lower
+    assert thinning >= 1
+    collection_size = (upper - lower) // thinning if collection_size is None else collection_size // thinning
+    assert collection_size >= (upper - lower) // thinning
     init_val_flat, unravel_fn = ravel_pytree(transform(init_val))
+    start_idx = lower + (upper - lower) % thinning
 
     @cached_by(fori_collect, body_fun, transform)
     def _body_fn(i, vals):
-        val, collection, lower_idx = vals
+        val, collection, start_idx, thinning = vals
         val = body_fun(val)
-        i = jnp.where(i >= lower_idx, i - lower_idx, 0)
-        collection = ops.index_update(collection, i, ravel_pytree(transform(val))[0])
-        return val, collection, lower_idx
+        idx = (i - start_idx) // thinning
+        collection = ops.index_update(collection, idx, ravel_pytree(transform(val))[0])
```

Comment:
nit: how about setting
```python
collection = cond(idx >= 0,
                  collection,
                  lambda x: ops.index_update(x, idx, ravel_pytree(transform(val))[0]),
                  collection,
                  identity)
```
to only update collection when the `idx` is nonnegative.